src/Elm/Print.hs

```diff
@@ -421,26 +428,211 @@ typeRefEncoder (RefPrim elmPrim) = case elmPrim of
+       decoderDef elmTypeName elmTypeVars
+    <> line
+    <> if isEnum t
+       -- if this is Enum just using the show instance we wrote.
+       then enumDecoder
+       -- If it sum type then it should look like: @{"tag": "Foo", "contents" : ["string", 1]}@
+       else sumDecoder
+  where
+    name :: Doc ann
+    name = decoderName elmTypeName <+> equals
+
+    typeName :: Doc ann
+    typeName = pretty elmTypeName
+
+    enumDecoder :: Doc ann
+    enumDecoder = name <+> "elmStreetDecodeEnum read" <> typeName
+
+    sumDecoder :: Doc ann
+    sumDecoder = nest 4 $ vsep
+        [ name
+        , nest 4 (vsep $ ("let decide : String -> Decoder" <+> typeName) :
+            [ nest 4
+                ( vsep $ "decide x = case x of"
+                : map cases (toList elmTypeConstructors)
+               ++ ["c -> D.fail <|" <+> dquotes (typeName <+> "doesn't have such constructor: ") <+> "++ c"]
+                )
+            ])
+        , "in D.andThen decide (D.field \"tag\" D.string)"
+        ]
+
+    cases :: ElmConstructor -> Doc ann
+    cases ElmConstructor{..} = dquotes cName <+> arrow <+>
+        case length elmConstructorFields of
+            0 -> "D.succeed" <+> cName
+            n -> "D.field \"contents\" <| D.map" <> mapNum n <+> cName <+> createIndexes
+
+      where
+        cName :: Doc ann
+        cName = pretty elmConstructorName
+
+        -- Use function map, map2, map3 etc.
+        mapNum :: Int -> Doc ann
+        mapNum 1 = emptyDoc
+        mapNum i = pretty i
+
+        createIndexes :: Doc ann
+        createIndexes = concatWith (surround " ") $ zipWith oneField [0..] elmConstructorFields
+
+        -- create @(D.index 0 D.string)@ etc.
+        oneField :: Int -> TypeRef -> Doc ann
+        oneField i typeRef = parens $ "D.index" <+> pretty i <+> typeRefDecoder typeRef
+
+
+-- | Converts the reference to the existing type to the corresponding decoder.
+typeRefDecoder :: TypeRef -> Doc ann
+typeRefDecoder (RefPrim elmPrim) = case elmPrim of
+    ElmUnit       -> "D.hardcoded ()"
+    ElmNever      -> "(D.fail \"Never is not possible\")"
+    ElmBool       -> "D.bool"
+    ElmChar       -> "elmStreetDecodeChar"
+    ElmInt        -> "D.int"
+    ElmFloat      -> "D.float"
+    ElmString     -> "D.string"
+    ElmMaybe t    -> parens $ "nullable" <+> typeRefEncoder t
+    ElmResult l r -> parens $ "elmStreetDecodeEither" <+> typeRefEncoder l <+> typeRefEncoder r
+    ElmPair a b   -> parens $ "elmStreetDecodePair" <+> typeRefEncoder a <+> typeRefEncoder b
+    ElmList l     -> parens $ "D.list" <+> typeRefEncoder l
+typeRefDecoder (RefCustom TypeName{..}) = "decode" <> pretty unTypeName
+
+-- | The definition of the @encodeTYPENAME@ function.
+decoderDef
+    :: Text  -- ^ Type name
+    -> [Text] -- ^ List of type variables
+    -> Doc ann
+decoderDef typeName vars =
+    decoderName typeName <+> colon <+> "Decoder" <+> typeVarsDoc
+  where
+    typeVarsDoc :: Doc ann
+    typeVarsDoc = case vars of
```

Comment:
You can use `typeWithVarsDoc` helper function here instead of having one inside `where`